{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "patient-congress",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>qname</th>\n",
       "      <th>flag</th>\n",
       "      <th>rname</th>\n",
       "      <th>pos</th>\n",
       "      <th>mapq</th>\n",
       "      <th>cigar</th>\n",
       "      <th>rnext</th>\n",
       "      <th>pnext</th>\n",
       "      <th>tlen</th>\n",
       "      <th>seq</th>\n",
       "      <th>qual</th>\n",
       "      <th>opt1</th>\n",
       "      <th>opt2</th>\n",
       "      <th>opt3</th>\n",
       "      <th>opt4</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SRR458756.2.803</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSMUST00000240377.1|ENSMUSG00000119584.1|OTTM...</td>\n",
       "      <td>295</td>\n",
       "      <td>255</td>\n",
       "      <td>15M</td>\n",
       "      <td>*</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>TGGTGACTCTAGATA</td>\n",
       "      <td>B@@@@AAAAAAAAAA</td>\n",
       "      <td>NH:i:1</td>\n",
       "      <td>HI:i:1</td>\n",
       "      <td>AS:i:14</td>\n",
       "      <td>nM:i:0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SRR458756.2.918</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSMUST00000091752.5|ENSMUSG00000069310.5|OTTM...</td>\n",
       "      <td>100</td>\n",
       "      <td>255</td>\n",
       "      <td>30M</td>\n",
       "      <td>*</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>CCAAGAATGGCTGGTACTAAGCAGACCGCT</td>\n",
       "      <td>A@A@@@A@@@?*.56A&lt;2A&lt;+;?*A??*?8</td>\n",
       "      <td>NH:i:1</td>\n",
       "      <td>HI:i:1</td>\n",
       "      <td>AS:i:27</td>\n",
       "      <td>nM:i:1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SRR458756.2.919</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSMUST00000240377.1|ENSMUSG00000119584.1|OTTM...</td>\n",
       "      <td>815</td>\n",
       "      <td>255</td>\n",
       "      <td>28M</td>\n",
       "      <td>*</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>AAAAAATTAGAGTGTTCAAAGCAGGCCC</td>\n",
       "      <td>IIIIIIIHIHGGBG8AFFFFDGGBBGGG</td>\n",
       "      <td>NH:i:1</td>\n",
       "      <td>HI:i:1</td>\n",
       "      <td>AS:i:27</td>\n",
       "      <td>nM:i:0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SRR458756.2.922</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSMUST00000240377.1|ENSMUSG00000119584.1|OTTM...</td>\n",
       "      <td>292</td>\n",
       "      <td>255</td>\n",
       "      <td>30M</td>\n",
       "      <td>*</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>CTTTGGCGACTCTAGATAACCTCGGGCCGA</td>\n",
       "      <td>##############################</td>\n",
       "      <td>NH:i:1</td>\n",
       "      <td>HI:i:1</td>\n",
       "      <td>AS:i:27</td>\n",
       "      <td>nM:i:1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SRR458756.2.923</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSMUST00000023741.16|ENSMUSG00000048154.18|OT...</td>\n",
       "      <td>4117</td>\n",
       "      <td>3</td>\n",
       "      <td>1S17M</td>\n",
       "      <td>*</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>AGTGGCTGATATTGATAG</td>\n",
       "      <td>FC?BCGGGBBGEG@GDG&gt;</td>\n",
       "      <td>NH:i:2</td>\n",
       "      <td>HI:i:1</td>\n",
       "      <td>AS:i:16</td>\n",
       "      <td>nM:i:0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             qname  flag                                              rname  \\\n",
       "0  SRR458756.2.803     0  ENSMUST00000240377.1|ENSMUSG00000119584.1|OTTM...   \n",
       "1  SRR458756.2.918     0  ENSMUST00000091752.5|ENSMUSG00000069310.5|OTTM...   \n",
       "2  SRR458756.2.919     0  ENSMUST00000240377.1|ENSMUSG00000119584.1|OTTM...   \n",
       "3  SRR458756.2.922     0  ENSMUST00000240377.1|ENSMUSG00000119584.1|OTTM...   \n",
       "4  SRR458756.2.923     0  ENSMUST00000023741.16|ENSMUSG00000048154.18|OT...   \n",
       "\n",
       "    pos  mapq  cigar rnext  pnext  tlen                             seq  \\\n",
       "0   295   255    15M     *      0     0                 TGGTGACTCTAGATA   \n",
       "1   100   255    30M     *      0     0  CCAAGAATGGCTGGTACTAAGCAGACCGCT   \n",
       "2   815   255    28M     *      0     0    AAAAAATTAGAGTGTTCAAAGCAGGCCC   \n",
       "3   292   255    30M     *      0     0  CTTTGGCGACTCTAGATAACCTCGGGCCGA   \n",
       "4  4117     3  1S17M     *      0     0              AGTGGCTGATATTGATAG   \n",
       "\n",
       "                             qual    opt1    opt2     opt3    opt4  \n",
       "0                 B@@@@AAAAAAAAAA  NH:i:1  HI:i:1  AS:i:14  nM:i:0  \n",
       "1  A@A@@@A@@@?*.56A<2A<+;?*A??*?8  NH:i:1  HI:i:1  AS:i:27  nM:i:1  \n",
       "2    IIIIIIIHIHGGBG8AFFFFDGGBBGGG  NH:i:1  HI:i:1  AS:i:27  nM:i:0  \n",
       "3  ##############################  NH:i:1  HI:i:1  AS:i:27  nM:i:1  \n",
       "4              FC?BCGGGBBGEG@GDG>  NH:i:2  HI:i:1  AS:i:16  nM:i:0  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import numpy as np\n",
    "import pandas as pd \n",
    "\n",
    "samData = pd.read_csv(\"/qbio/nest/prom317/tutorial/transcriptomeAlignment/RPF_siLuc/ForwardStrand.out.sam\", sep = '\\t', names=['qname', 'flag', 'rname', 'pos', 'mapq', 'cigar', 'rnext', 'pnext', 'tlen', 'seq', 'qual','opt1', 'opt2', 'opt3', 'opt4'])\n",
    "samData.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "civilian-fantasy",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(43076098, 17)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>qname</th>\n",
       "      <th>flag</th>\n",
       "      <th>rname</th>\n",
       "      <th>pos</th>\n",
       "      <th>mapq</th>\n",
       "      <th>cigar</th>\n",
       "      <th>rnext</th>\n",
       "      <th>pnext</th>\n",
       "      <th>tlen</th>\n",
       "      <th>seq</th>\n",
       "      <th>qual</th>\n",
       "      <th>opt1</th>\n",
       "      <th>opt2</th>\n",
       "      <th>opt3</th>\n",
       "      <th>opt4</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SRR458756.2.803</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSMUST00000240377.1|ENSMUSG00000119584.1|OTTM...</td>\n",
       "      <td>295</td>\n",
       "      <td>255</td>\n",
       "      <td>15M</td>\n",
       "      <td>*</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>TGGTGACTCTAGATA</td>\n",
       "      <td>B@@@@AAAAAAAAAA</td>\n",
       "      <td>NH:i:1</td>\n",
       "      <td>HI:i:1</td>\n",
       "      <td>AS:i:14</td>\n",
       "      <td>nM:i:0</td>\n",
       "      <td>ENSMUST00000240377.1</td>\n",
       "      <td>ENSMUSG00000119584.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SRR458756.2.918</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSMUST00000091752.5|ENSMUSG00000069310.5|OTTM...</td>\n",
       "      <td>100</td>\n",
       "      <td>255</td>\n",
       "      <td>30M</td>\n",
       "      <td>*</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>CCAAGAATGGCTGGTACTAAGCAGACCGCT</td>\n",
       "      <td>A@A@@@A@@@?*.56A&lt;2A&lt;+;?*A??*?8</td>\n",
       "      <td>NH:i:1</td>\n",
       "      <td>HI:i:1</td>\n",
       "      <td>AS:i:27</td>\n",
       "      <td>nM:i:1</td>\n",
       "      <td>ENSMUST00000091752.5</td>\n",
       "      <td>ENSMUSG00000069310.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SRR458756.2.919</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSMUST00000240377.1|ENSMUSG00000119584.1|OTTM...</td>\n",
       "      <td>815</td>\n",
       "      <td>255</td>\n",
       "      <td>28M</td>\n",
       "      <td>*</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>AAAAAATTAGAGTGTTCAAAGCAGGCCC</td>\n",
       "      <td>IIIIIIIHIHGGBG8AFFFFDGGBBGGG</td>\n",
       "      <td>NH:i:1</td>\n",
       "      <td>HI:i:1</td>\n",
       "      <td>AS:i:27</td>\n",
       "      <td>nM:i:0</td>\n",
       "      <td>ENSMUST00000240377.1</td>\n",
       "      <td>ENSMUSG00000119584.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SRR458756.2.922</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSMUST00000240377.1|ENSMUSG00000119584.1|OTTM...</td>\n",
       "      <td>292</td>\n",
       "      <td>255</td>\n",
       "      <td>30M</td>\n",
       "      <td>*</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>CTTTGGCGACTCTAGATAACCTCGGGCCGA</td>\n",
       "      <td>##############################</td>\n",
       "      <td>NH:i:1</td>\n",
       "      <td>HI:i:1</td>\n",
       "      <td>AS:i:27</td>\n",
       "      <td>nM:i:1</td>\n",
       "      <td>ENSMUST00000240377.1</td>\n",
       "      <td>ENSMUSG00000119584.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SRR458756.2.923</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSMUST00000023741.16|ENSMUSG00000048154.18|OT...</td>\n",
       "      <td>4117</td>\n",
       "      <td>3</td>\n",
       "      <td>1S17M</td>\n",
       "      <td>*</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>AGTGGCTGATATTGATAG</td>\n",
       "      <td>FC?BCGGGBBGEG@GDG&gt;</td>\n",
       "      <td>NH:i:2</td>\n",
       "      <td>HI:i:1</td>\n",
       "      <td>AS:i:16</td>\n",
       "      <td>nM:i:0</td>\n",
       "      <td>ENSMUST00000023741.16</td>\n",
       "      <td>ENSMUSG00000048154.18</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             qname  flag                                              rname  \\\n",
       "0  SRR458756.2.803     0  ENSMUST00000240377.1|ENSMUSG00000119584.1|OTTM...   \n",
       "1  SRR458756.2.918     0  ENSMUST00000091752.5|ENSMUSG00000069310.5|OTTM...   \n",
       "2  SRR458756.2.919     0  ENSMUST00000240377.1|ENSMUSG00000119584.1|OTTM...   \n",
       "3  SRR458756.2.922     0  ENSMUST00000240377.1|ENSMUSG00000119584.1|OTTM...   \n",
       "4  SRR458756.2.923     0  ENSMUST00000023741.16|ENSMUSG00000048154.18|OT...   \n",
       "\n",
       "    pos  mapq  cigar rnext  pnext  tlen                             seq  \\\n",
       "0   295   255    15M     *      0     0                 TGGTGACTCTAGATA   \n",
       "1   100   255    30M     *      0     0  CCAAGAATGGCTGGTACTAAGCAGACCGCT   \n",
       "2   815   255    28M     *      0     0    AAAAAATTAGAGTGTTCAAAGCAGGCCC   \n",
       "3   292   255    30M     *      0     0  CTTTGGCGACTCTAGATAACCTCGGGCCGA   \n",
       "4  4117     3  1S17M     *      0     0              AGTGGCTGATATTGATAG   \n",
       "\n",
       "                             qual    opt1    opt2     opt3    opt4  \\\n",
       "0                 B@@@@AAAAAAAAAA  NH:i:1  HI:i:1  AS:i:14  nM:i:0   \n",
       "1  A@A@@@A@@@?*.56A<2A<+;?*A??*?8  NH:i:1  HI:i:1  AS:i:27  nM:i:1   \n",
       "2    IIIIIIIHIHGGBG8AFFFFDGGBBGGG  NH:i:1  HI:i:1  AS:i:27  nM:i:0   \n",
       "3  ##############################  NH:i:1  HI:i:1  AS:i:27  nM:i:1   \n",
       "4              FC?BCGGGBBGEG@GDG>  NH:i:2  HI:i:1  AS:i:16  nM:i:0   \n",
       "\n",
       "           transcript_id                gene_id  \n",
       "0   ENSMUST00000240377.1   ENSMUSG00000119584.1  \n",
       "1   ENSMUST00000091752.5   ENSMUSG00000069310.5  \n",
       "2   ENSMUST00000240377.1   ENSMUSG00000119584.1  \n",
       "3   ENSMUST00000240377.1   ENSMUSG00000119584.1  \n",
       "4  ENSMUST00000023741.16  ENSMUSG00000048154.18  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "samData['transcript_id'] = samData.rname.map(lambda p : p.split('|')[0])\n",
    "samData['gene_id'] = samData.rname.map(lambda p : p.split('|')[1])\n",
    "print(samData.shape)\n",
    "samData.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "prostate-sailing",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seq_name</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attribute</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>3143476.0</td>\n",
       "      <td>3144545.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000102693.2\"; gene_type \"TEC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>3143476.0</td>\n",
       "      <td>3144545.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000102693.2\"; transcript_id ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>3143476.0</td>\n",
       "      <td>3144545.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000102693.2\"; transcript_id ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>gene</td>\n",
       "      <td>3172239.0</td>\n",
       "      <td>3172348.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000064842.3\"; gene_type \"snR...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>transcript</td>\n",
       "      <td>3172239.0</td>\n",
       "      <td>3172348.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000064842.3\"; transcript_id ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  seq_name   source     feature      start        end score strand frame  \\\n",
       "5     chr1   HAVANA        gene  3143476.0  3144545.0     .      +     .   \n",
       "6     chr1   HAVANA  transcript  3143476.0  3144545.0     .      +     .   \n",
       "7     chr1   HAVANA        exon  3143476.0  3144545.0     .      +     .   \n",
       "8     chr1  ENSEMBL        gene  3172239.0  3172348.0     .      +     .   \n",
       "9     chr1  ENSEMBL  transcript  3172239.0  3172348.0     .      +     .   \n",
       "\n",
       "                                           attribute  \n",
       "5  gene_id \"ENSMUSG00000102693.2\"; gene_type \"TEC...  \n",
       "6  gene_id \"ENSMUSG00000102693.2\"; transcript_id ...  \n",
       "7  gene_id \"ENSMUSG00000102693.2\"; transcript_id ...  \n",
       "8  gene_id \"ENSMUSG00000064842.3\"; gene_type \"snR...  \n",
       "9  gene_id \"ENSMUSG00000064842.3\"; transcript_id ...  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gtfData = pd.read_table(\"/qbio/nest/prom317/tutorial/references/GRCm39.genome.gtf\", sep = '\\t', names = [\"seq_name\",\"source\",\"feature\",\"start\",\"end\",\"score\",\"strand\",\"frame\",\"attribute\"])\n",
    "gtfData.iloc[5:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "every-muscle",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>qname</th>\n",
       "      <th>flag</th>\n",
       "      <th>rname</th>\n",
       "      <th>pos</th>\n",
       "      <th>mapq</th>\n",
       "      <th>cigar</th>\n",
       "      <th>rnext</th>\n",
       "      <th>pnext</th>\n",
       "      <th>tlen</th>\n",
       "      <th>seq</th>\n",
       "      <th>qual</th>\n",
       "      <th>opt1</th>\n",
       "      <th>opt2</th>\n",
       "      <th>opt3</th>\n",
       "      <th>opt4</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>start_pos</th>\n",
       "      <th>end_pos</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SRR458756.2.803</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSMUST00000240377.1|ENSMUSG00000119584.1|OTTM...</td>\n",
       "      <td>295</td>\n",
       "      <td>255</td>\n",
       "      <td>15M</td>\n",
       "      <td>*</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>TGGTGACTCTAGATA</td>\n",
       "      <td>B@@@@AAAAAAAAAA</td>\n",
       "      <td>NH:i:1</td>\n",
       "      <td>HI:i:1</td>\n",
       "      <td>AS:i:14</td>\n",
       "      <td>nM:i:0</td>\n",
       "      <td>ENSMUST00000240377.1</td>\n",
       "      <td>ENSMUSG00000119584.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SRR458756.2.918</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSMUST00000091752.5|ENSMUSG00000069310.5|OTTM...</td>\n",
       "      <td>100</td>\n",
       "      <td>255</td>\n",
       "      <td>30M</td>\n",
       "      <td>*</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>CCAAGAATGGCTGGTACTAAGCAGACCGCT</td>\n",
       "      <td>A@A@@@A@@@?*.56A&lt;2A&lt;+;?*A??*?8</td>\n",
       "      <td>NH:i:1</td>\n",
       "      <td>HI:i:1</td>\n",
       "      <td>AS:i:27</td>\n",
       "      <td>nM:i:1</td>\n",
       "      <td>ENSMUST00000091752.5</td>\n",
       "      <td>ENSMUSG00000069310.5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SRR458756.2.919</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSMUST00000240377.1|ENSMUSG00000119584.1|OTTM...</td>\n",
       "      <td>815</td>\n",
       "      <td>255</td>\n",
       "      <td>28M</td>\n",
       "      <td>*</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>AAAAAATTAGAGTGTTCAAAGCAGGCCC</td>\n",
       "      <td>IIIIIIIHIHGGBG8AFFFFDGGBBGGG</td>\n",
       "      <td>NH:i:1</td>\n",
       "      <td>HI:i:1</td>\n",
       "      <td>AS:i:27</td>\n",
       "      <td>nM:i:0</td>\n",
       "      <td>ENSMUST00000240377.1</td>\n",
       "      <td>ENSMUSG00000119584.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SRR458756.2.922</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSMUST00000240377.1|ENSMUSG00000119584.1|OTTM...</td>\n",
       "      <td>292</td>\n",
       "      <td>255</td>\n",
       "      <td>30M</td>\n",
       "      <td>*</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>CTTTGGCGACTCTAGATAACCTCGGGCCGA</td>\n",
       "      <td>##############################</td>\n",
       "      <td>NH:i:1</td>\n",
       "      <td>HI:i:1</td>\n",
       "      <td>AS:i:27</td>\n",
       "      <td>nM:i:1</td>\n",
       "      <td>ENSMUST00000240377.1</td>\n",
       "      <td>ENSMUSG00000119584.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SRR458756.2.923</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSMUST00000023741.16|ENSMUSG00000048154.18|OT...</td>\n",
       "      <td>4117</td>\n",
       "      <td>3</td>\n",
       "      <td>1S17M</td>\n",
       "      <td>*</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>AGTGGCTGATATTGATAG</td>\n",
       "      <td>FC?BCGGGBBGEG@GDG&gt;</td>\n",
       "      <td>NH:i:2</td>\n",
       "      <td>HI:i:1</td>\n",
       "      <td>AS:i:16</td>\n",
       "      <td>nM:i:0</td>\n",
       "      <td>ENSMUST00000023741.16</td>\n",
       "      <td>ENSMUSG00000048154.18</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             qname  flag                                              rname  \\\n",
       "0  SRR458756.2.803     0  ENSMUST00000240377.1|ENSMUSG00000119584.1|OTTM...   \n",
       "1  SRR458756.2.918     0  ENSMUST00000091752.5|ENSMUSG00000069310.5|OTTM...   \n",
       "2  SRR458756.2.919     0  ENSMUST00000240377.1|ENSMUSG00000119584.1|OTTM...   \n",
       "3  SRR458756.2.922     0  ENSMUST00000240377.1|ENSMUSG00000119584.1|OTTM...   \n",
       "4  SRR458756.2.923     0  ENSMUST00000023741.16|ENSMUSG00000048154.18|OT...   \n",
       "\n",
       "    pos  mapq  cigar rnext  pnext  tlen                             seq  \\\n",
       "0   295   255    15M     *      0     0                 TGGTGACTCTAGATA   \n",
       "1   100   255    30M     *      0     0  CCAAGAATGGCTGGTACTAAGCAGACCGCT   \n",
       "2   815   255    28M     *      0     0    AAAAAATTAGAGTGTTCAAAGCAGGCCC   \n",
       "3   292   255    30M     *      0     0  CTTTGGCGACTCTAGATAACCTCGGGCCGA   \n",
       "4  4117     3  1S17M     *      0     0              AGTGGCTGATATTGATAG   \n",
       "\n",
       "                             qual    opt1    opt2     opt3    opt4  \\\n",
       "0                 B@@@@AAAAAAAAAA  NH:i:1  HI:i:1  AS:i:14  nM:i:0   \n",
       "1  A@A@@@A@@@?*.56A<2A<+;?*A??*?8  NH:i:1  HI:i:1  AS:i:27  nM:i:1   \n",
       "2    IIIIIIIHIHGGBG8AFFFFDGGBBGGG  NH:i:1  HI:i:1  AS:i:27  nM:i:0   \n",
       "3  ##############################  NH:i:1  HI:i:1  AS:i:27  nM:i:1   \n",
       "4              FC?BCGGGBBGEG@GDG>  NH:i:2  HI:i:1  AS:i:16  nM:i:0   \n",
       "\n",
       "           transcript_id                gene_id  start_pos  end_pos  \n",
       "0   ENSMUST00000240377.1   ENSMUSG00000119584.1        NaN      NaN  \n",
       "1   ENSMUST00000091752.5   ENSMUSG00000069310.5        NaN      NaN  \n",
       "2   ENSMUST00000240377.1   ENSMUSG00000119584.1        NaN      NaN  \n",
       "3   ENSMUST00000240377.1   ENSMUSG00000119584.1        NaN      NaN  \n",
       "4  ENSMUST00000023741.16  ENSMUSG00000048154.18        NaN      NaN  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "samData['start_pos'] = np.nan\n",
    "samData['end_pos'] = np.nan\n",
    "samData.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "unnecessary-artist",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "ENSMUSG00000052613.17    2003\n",
       "ENSMUSG00000069601.15    1772\n",
       "ENSMUSG00000021843.18    1695\n",
       "ENSMUSG00000033788.16    1411\n",
       "ENSMUSG00000064302.14    1411\n",
       "                         ... \n",
       "ENSMUSG00000107199.2        3\n",
       "ENSMUSG00000098293.3        3\n",
       "ENSMUSG00000109837.2        3\n",
       "ENSMUSG00000077603.3        3\n",
       "ENSMUSG00000095779.3        3\n",
       "Name: gene_id, Length: 26876, dtype: int64"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gtfPlusData = gtfData.loc[gtfData.strand == \"+\"]\n",
    "del gtfData\n",
    "gtfPlusData['gene_id'] = gtfPlusData.attribute.map(lambda p: p.split(';')[0].split(' ')[1][1:-1])\n",
    "gtfPlusData.gene_id.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "mediterranean-bridges",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['ENSMUST00000240377.1' 'ENSMUST00000091752.5' 'ENSMUST00000023741.16' ...\n",
      " 'ENSMUST00000141667.2' 'ENSMUST00000099986.3' 'ENSMUST00000225350.2']\n"
     ]
    }
   ],
   "source": [
    "transcript_list = samData['transcript_id'].unique()\n",
    "transDict = {}\n",
    "for ts in transcript_list:\n",
    "    transDict[ts] = [[],[]]\n",
    "print(transcript_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "known-bleeding",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/blaze/prom317/conda/envs/lab/lib/python3.7/site-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  \n",
      "/blaze/prom317/conda/envs/lab/lib/python3.7/site-packages/ipykernel_launcher.py:4: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  after removing the cwd from sys.path.\n",
      "/blaze/prom317/conda/envs/lab/lib/python3.7/site-packages/ipykernel_launcher.py:6: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  \n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seq_name</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attribute</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>361</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>4878137.0</td>\n",
       "      <td>4878139.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"ENSMUSG00000025903.15\"; transcript_id...</td>\n",
       "      <td>ENSMUSG00000025903.15</td>\n",
       "      <td>ENSMUST00000134384.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>386</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>4878137.0</td>\n",
       "      <td>4878139.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"ENSMUSG00000025903.15\"; transcript_id...</td>\n",
       "      <td>ENSMUSG00000025903.15</td>\n",
       "      <td>ENSMUST00000027036.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>409</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>4878137.0</td>\n",
       "      <td>4878139.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"ENSMUSG00000025903.15\"; transcript_id...</td>\n",
       "      <td>ENSMUSG00000025903.15</td>\n",
       "      <td>ENSMUST00000150971.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>430</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>4878137.0</td>\n",
       "      <td>4878139.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"ENSMUSG00000025903.15\"; transcript_id...</td>\n",
       "      <td>ENSMUSG00000025903.15</td>\n",
       "      <td>ENSMUST00000119612.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>449</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>4878137.0</td>\n",
       "      <td>4878139.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"ENSMUSG00000025903.15\"; transcript_id...</td>\n",
       "      <td>ENSMUSG00000025903.15</td>\n",
       "      <td>ENSMUST00000137887.8</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    seq_name  source      feature      start        end score strand frame  \\\n",
       "361     chr1  HAVANA  start_codon  4878137.0  4878139.0     .      +     0   \n",
       "386     chr1  HAVANA  start_codon  4878137.0  4878139.0     .      +     0   \n",
       "409     chr1  HAVANA  start_codon  4878137.0  4878139.0     .      +     0   \n",
       "430     chr1  HAVANA  start_codon  4878137.0  4878139.0     .      +     0   \n",
       "449     chr1  HAVANA  start_codon  4878137.0  4878139.0     .      +     0   \n",
       "\n",
       "                                             attribute                gene_id  \\\n",
       "361  gene_id \"ENSMUSG00000025903.15\"; transcript_id...  ENSMUSG00000025903.15   \n",
       "386  gene_id \"ENSMUSG00000025903.15\"; transcript_id...  ENSMUSG00000025903.15   \n",
       "409  gene_id \"ENSMUSG00000025903.15\"; transcript_id...  ENSMUSG00000025903.15   \n",
       "430  gene_id \"ENSMUSG00000025903.15\"; transcript_id...  ENSMUSG00000025903.15   \n",
       "449  gene_id \"ENSMUSG00000025903.15\"; transcript_id...  ENSMUSG00000025903.15   \n",
       "\n",
       "             transcript_id  \n",
       "361   ENSMUST00000134384.8  \n",
       "386  ENSMUST00000027036.11  \n",
       "409   ENSMUST00000150971.8  \n",
       "430   ENSMUST00000119612.9  \n",
       "449   ENSMUST00000137887.8  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gtfExonData = gtfPlusData.loc[gtfPlusData.feature == 'exon']\n",
    "gtfExonData[\"transcript_id\"] = gtfExonData.iloc[:,8].map(lambda p: p.split(';')[1].split(' ')[2][1:-1])\n",
    "gtfStartData = gtfPlusData.loc[gtfPlusData.feature == 'start_codon']\n",
    "gtfStartData[\"transcript_id\"] = gtfStartData.iloc[:,8].map(lambda p: p.split(';')[1].split(' ')[2][1:-1])\n",
    "gtfStopData = gtfPlusData.loc[gtfPlusData.feature == 'stop_codon']\n",
    "gtfStopData[\"transcript_id\"] = gtfStopData.iloc[:,8].map(lambda p: p.split(';')[1].split(' ')[2][1:-1])\n",
    "del gtfPlusData\n",
    "gtfStartData.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "intimate-gateway",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1000\n",
      "2000\n",
      "3000\n",
      "4000\n",
      "5000\n",
      "6000\n",
      "7000\n",
      "8000\n",
      "9000\n",
      "10000\n",
      "11000\n",
      "12000\n",
      "13000\n",
      "14000\n",
      "15000\n",
      "16000\n",
      "17000\n",
      "18000\n",
      "19000\n",
      "20000\n",
      "21000\n",
      "22000\n",
      "23000\n",
      "24000\n",
      "25000\n",
      "26000\n",
      "27000\n",
      "28000\n",
      "29000\n",
      "30000\n",
      "31000\n",
      "32000\n",
      "33000\n",
      "34000\n",
      "35000\n",
      "36000\n",
      "37000\n",
      "38000\n",
      "39000\n",
      "40000\n",
      "41000\n",
      "42000\n",
      "43000\n",
      "44000\n",
      "45000\n",
      "46000\n",
      "47000\n",
      "48000\n",
      "49000\n",
      "50000\n",
      "51000\n",
      "52000\n",
      "53000\n",
      "54000\n",
      "55000\n",
      "56000\n",
      "57000\n",
      "58000\n",
      "59000\n",
      "60000\n",
      "61000\n",
      "62000\n",
      "63000\n",
      "64000\n",
      "65000\n",
      "66000\n",
      "67000\n",
      "68000\n",
      "69000\n",
      "70000\n",
      "71000\n",
      "72000\n",
      "73000\n",
      "74000\n",
      "75000\n",
      "76000\n",
      "77000\n",
      "78000\n",
      "79000\n",
      "80000\n",
      "81000\n",
      "82000\n",
      "83000\n",
      "84000\n",
      "85000\n",
      "86000\n",
      "87000\n",
      "88000\n",
      "89000\n",
      "90000\n",
      "91000\n",
      "92000\n",
      "93000\n",
      "94000\n",
      "95000\n",
      "96000\n",
      "97000\n",
      "98000\n",
      "99000\n",
      "100000\n",
      "101000\n",
      "102000\n",
      "103000\n",
      "104000\n",
      "105000\n",
      "106000\n"
     ]
    }
   ],
   "source": [
    "count = 0\n",
    "for ts in transDict.keys():\n",
    "    count += 1\n",
    "    if count % 1000 == 0:\n",
    "        print(count)\n",
    "    tsGtfExon = gtfExonData.loc[gtfExonData.transcript_id == ts]\n",
    "    tsGtfStart = gtfStartData.loc[gtfStartData.transcript_id == ts]\n",
    "    tsGtfStop = gtfStopData.loc[gtfStopData.transcript_id == ts]\n",
    "    for idx, row in tsGtfExon.iterrows():\n",
    "        transDict[ts][0].append(row['start'])\n",
    "        transDict[ts][1].append(row['end'])\n",
    "    if not (tsGtfStart.empty):\n",
    "        transDict[ts].append(['start_codon',tsGtfStart.iloc[0,3]] )\n",
    "    if not (tsGtfStop.empty):\n",
    "        transDict[ts].append(['stop_codon',tsGtfStop.iloc[0,3]] )\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "grateful-singles",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "21605\n"
     ]
    }
   ],
   "source": [
    "codingTransDict = {}\n",
    "for i in transDict.keys():\n",
    "    if len(transDict[i]) == 4:\n",
    "        codingTransDict[i] = transDict[i]\n",
    "print(len(codingTransDict))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "induced-accessory",
   "metadata": {},
   "outputs": [],
   "source": [
    "for i in codingTransDict.keys():\n",
    "    exonStart = codingTransDict[i][0]\n",
    "    exonEnd = codingTransDict[i][1]\n",
    "    startCod = codingTransDict[i][2][1]\n",
    "    stopCod = codingTransDict[i][3][1]\n",
    "    exonLength = 0\n",
    "    startPos = 0\n",
    "    stopPos = 0\n",
    "    for j in range(len(exonStart)):\n",
    "        if exonStart[j] <= startCod < exonEnd[j]:\n",
    "            startPos = exonLength + startCod - exonStart[j] + 1\n",
    "            exonLength = exonLength + exonEnd[j] - exonStart[j] + 1\n",
    "        elif exonStart[j] <= stopCod < exonEnd[j]:\n",
    "            stopPos = exonLength + stopCod - exonStart[j] + 1\n",
    "            exonLength = exonLength + exonEnd[j] - exonStart[j] + 1\n",
    "        else:\n",
    "            exonLength = exonLength + exonEnd[j] - exonStart[j] + 1\n",
    "    codingTransDict[i].append(exonLength)\n",
    "    codingTransDict[i].append(startPos)\n",
    "    codingTransDict[i].append(stopPos)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "given-wealth",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSMUST00000057000.17</th>\n",
       "      <td>[98713242.0, 98713739.0, 98719037.0, 98722444....</td>\n",
       "      <td>[98713417.0, 98713781.0, 98719151.0, 98722484....</td>\n",
       "      <td>[start_codon, 98713382.0]</td>\n",
       "      <td>[stop_codon, 98752838.0]</td>\n",
       "      <td>3046.0</td>\n",
       "      <td>141.0</td>\n",
       "      <td>1878.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUST00000110105.10</th>\n",
       "      <td>[98713248.0, 98713739.0, 98719037.0, 98722444....</td>\n",
       "      <td>[98713417.0, 98713781.0, 98719151.0, 98722484....</td>\n",
       "      <td>[start_codon, 98713382.0]</td>\n",
       "      <td>[stop_codon, 98752838.0]</td>\n",
       "      <td>3514.0</td>\n",
       "      <td>135.0</td>\n",
       "      <td>2340.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUST00000110104.10</th>\n",
       "      <td>[98713223.0, 98713739.0, 98719037.0, 98722444....</td>\n",
       "      <td>[98713417.0, 98713781.0, 98719151.0, 98722484....</td>\n",
       "      <td>[start_codon, 98713382.0]</td>\n",
       "      <td>[stop_codon, 98752838.0]</td>\n",
       "      <td>3146.0</td>\n",
       "      <td>160.0</td>\n",
       "      <td>1972.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUST00000067664.10</th>\n",
       "      <td>[75623695.0, 75642707.0, 75646400.0, 75647643....</td>\n",
       "      <td>[75623917.0, 75642906.0, 75646506.0, 75647849....</td>\n",
       "      <td>[start_codon, 75623854.0]</td>\n",
       "      <td>[stop_codon, 75655496.0]</td>\n",
       "      <td>2100.0</td>\n",
       "      <td>160.0</td>\n",
       "      <td>925.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUST00000041042.13</th>\n",
       "      <td>[4136789.0, 4155757.0, 4159926.0, 4162141.0, 4...</td>\n",
       "      <td>[4137099.0, 4155856.0, 4160033.0, 4162247.0, 4...</td>\n",
       "      <td>[start_codon, 4136789.0]</td>\n",
       "      <td>[stop_codon, 4165151.0]</td>\n",
       "      <td>1957.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1603.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                       0  \\\n",
       "ENSMUST00000057000.17  [98713242.0, 98713739.0, 98719037.0, 98722444....   \n",
       "ENSMUST00000110105.10  [98713248.0, 98713739.0, 98719037.0, 98722444....   \n",
       "ENSMUST00000110104.10  [98713223.0, 98713739.0, 98719037.0, 98722444....   \n",
       "ENSMUST00000067664.10  [75623695.0, 75642707.0, 75646400.0, 75647643....   \n",
       "ENSMUST00000041042.13  [4136789.0, 4155757.0, 4159926.0, 4162141.0, 4...   \n",
       "\n",
       "                                                                       1  \\\n",
       "ENSMUST00000057000.17  [98713417.0, 98713781.0, 98719151.0, 98722484....   \n",
       "ENSMUST00000110105.10  [98713417.0, 98713781.0, 98719151.0, 98722484....   \n",
       "ENSMUST00000110104.10  [98713417.0, 98713781.0, 98719151.0, 98722484....   \n",
       "ENSMUST00000067664.10  [75623917.0, 75642906.0, 75646506.0, 75647849....   \n",
       "ENSMUST00000041042.13  [4137099.0, 4155856.0, 4160033.0, 4162247.0, 4...   \n",
       "\n",
       "                                               2                         3  \\\n",
       "ENSMUST00000057000.17  [start_codon, 98713382.0]  [stop_codon, 98752838.0]   \n",
       "ENSMUST00000110105.10  [start_codon, 98713382.0]  [stop_codon, 98752838.0]   \n",
       "ENSMUST00000110104.10  [start_codon, 98713382.0]  [stop_codon, 98752838.0]   \n",
       "ENSMUST00000067664.10  [start_codon, 75623854.0]  [stop_codon, 75655496.0]   \n",
       "ENSMUST00000041042.13   [start_codon, 4136789.0]   [stop_codon, 4165151.0]   \n",
       "\n",
       "                            4      5       6  \n",
       "ENSMUST00000057000.17  3046.0  141.0  1878.0  \n",
       "ENSMUST00000110105.10  3514.0  135.0  2340.0  \n",
       "ENSMUST00000110104.10  3146.0  160.0  1972.0  \n",
       "ENSMUST00000067664.10  2100.0  160.0   925.0  \n",
       "ENSMUST00000041042.13  1957.0    1.0  1603.0  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "codingDF = pd.DataFrame.from_dict(codingTransDict, orient = 'index')\n",
    "codingDF.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "headed-marriage",
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-14-d646cba13a04>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      5\u001b[0m         \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcount\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      6\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0msamData\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0miloc\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mi\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;36m15\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mcodingTransDict\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mkeys\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 7\u001b[0;31m         \u001b[0msamData\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0miloc\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mi\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;36m17\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m/blaze/prom317/conda/envs/lab/lib/python3.7/site-packages/pandas/core/indexing.py\u001b[0m in \u001b[0;36m__setitem__\u001b[0;34m(self, key, value)\u001b[0m\n\u001b[1;32m    690\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    691\u001b[0m         \u001b[0miloc\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m \u001b[0;32mif\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mname\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;34m\"iloc\"\u001b[0m \u001b[0;32melse\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mobj\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0miloc\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 692\u001b[0;31m         \u001b[0miloc\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_setitem_with_indexer\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mindexer\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvalue\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mname\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    693\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    694\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m_validate_key\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mkey\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0maxis\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mint\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/blaze/prom317/conda/envs/lab/lib/python3.7/site-packages/pandas/core/indexing.py\u001b[0m in \u001b[0;36m_setitem_with_indexer\u001b[0;34m(self, indexer, value, name)\u001b[0m\n\u001b[1;32m   1633\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mtake_split_path\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1634\u001b[0m             \u001b[0;31m# We have to operate column-wise\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1635\u001b[0;31m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_setitem_with_indexer_split_path\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mindexer\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvalue\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mname\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1636\u001b[0m         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1637\u001b[0m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_setitem_single_block\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mindexer\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvalue\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mname\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/blaze/prom317/conda/envs/lab/lib/python3.7/site-packages/pandas/core/indexing.py\u001b[0m in \u001b[0;36m_setitem_with_indexer_split_path\u001b[0;34m(self, indexer, value, name)\u001b[0m\n\u001b[1;32m   1718\u001b[0m             \u001b[0;31m# scalar value\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1719\u001b[0m             \u001b[0;32mfor\u001b[0m \u001b[0mloc\u001b[0m \u001b[0;32min\u001b[0m \u001b[0milocs\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1720\u001b[0;31m                 \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_setitem_single_column\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mloc\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvalue\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mpi\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1721\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1722\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m_setitem_with_indexer_2d_value\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mindexer\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvalue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/blaze/prom317/conda/envs/lab/lib/python3.7/site-packages/pandas/core/indexing.py\u001b[0m in \u001b[0;36m_setitem_single_column\u001b[0;34m(self, loc, value, plane_indexer)\u001b[0m\n\u001b[1;32m   1815\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1816\u001b[0m         \u001b[0;31m# reset the sliced object if unique\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1817\u001b[0;31m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mobj\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_iset_item\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mloc\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mser\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1818\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1819\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m_setitem_single_block\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mindexer\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvalue\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mname\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mstr\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/blaze/prom317/conda/envs/lab/lib/python3.7/site-packages/pandas/core/frame.py\u001b[0m in \u001b[0;36m_iset_item\u001b[0;34m(self, loc, value)\u001b[0m\n\u001b[1;32m   3220\u001b[0m         \u001b[0;31m# technically _sanitize_column expects a label, not a position,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3221\u001b[0m         \u001b[0;31m#  but the behavior is the same as long as we pass broadcast=False\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 3222\u001b[0;31m         \u001b[0mvalue\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_sanitize_column\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mloc\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvalue\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mbroadcast\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mFalse\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   3223\u001b[0m         \u001b[0mNDFrame\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_iset_item\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mloc\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvalue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3224\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/blaze/prom317/conda/envs/lab/lib/python3.7/site-packages/pandas/core/frame.py\u001b[0m in \u001b[0;36m_sanitize_column\u001b[0;34m(self, key, value, broadcast)\u001b[0m\n\u001b[1;32m   3874\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3875\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0misinstance\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mvalue\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mSeries\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 3876\u001b[0;31m             \u001b[0mvalue\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mreindexer\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mvalue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   3877\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3878\u001b[0m         \u001b[0;32melif\u001b[0m \u001b[0misinstance\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mvalue\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mDataFrame\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/blaze/prom317/conda/envs/lab/lib/python3.7/site-packages/pandas/core/frame.py\u001b[0m in \u001b[0;36mreindexer\u001b[0;34m(value)\u001b[0m\n\u001b[1;32m   3855\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3856\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mvalue\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mindex\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mequals\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mindex\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mor\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mindex\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 3857\u001b[0;31m                 \u001b[0mvalue\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mvalue\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_values\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcopy\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   3858\u001b[0m             \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3859\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "count = 0\n",
    "for i in range(43076098):\n",
    "    count +=1\n",
    "    if count % 100000 ==0:\n",
    "        print(count)\n",
    "    if samData.iloc[i,15] in codingTransDict.keys():\n",
    "        samData.iloc[i, 17] =1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "derived-radius",
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-24-28d92a4456ec>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0;32mfor\u001b[0m \u001b[0midx1\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mrow1\u001b[0m \u001b[0;32min\u001b[0m \u001b[0msamData\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0miterrows\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      2\u001b[0m     \u001b[0mcount\u001b[0m \u001b[0;34m+=\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      3\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0mcount\u001b[0m \u001b[0;34m%\u001b[0m \u001b[0;36m1000\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;36m0\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m         \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcount\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m     \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/blaze/prom317/conda/envs/lab/lib/python3.7/site-packages/pandas/core/frame.py\u001b[0m in \u001b[0;36miterrows\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m   1068\u001b[0m         \u001b[0mcolumns\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcolumns\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1069\u001b[0m         \u001b[0mklass\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_constructor_sliced\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1070\u001b[0;31m         \u001b[0;32mfor\u001b[0m \u001b[0mk\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mv\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mzip\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mindex\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mvalues\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1071\u001b[0m             \u001b[0ms\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mklass\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mv\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mindex\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mcolumns\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mname\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mk\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1072\u001b[0m             \u001b[0;32myield\u001b[0m \u001b[0mk\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0ms\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/blaze/prom317/conda/envs/lab/lib/python3.7/site-packages/pandas/core/generic.py\u001b[0m in \u001b[0;36mvalues\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m   5671\u001b[0m         \"\"\"\n\u001b[1;32m   5672\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_consolidate_inplace\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 5673\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_mgr\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mas_array\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtranspose\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_AXIS_REVERSED\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   5674\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   5675\u001b[0m     \u001b[0;34m@\u001b[0m\u001b[0mproperty\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/blaze/prom317/conda/envs/lab/lib/python3.7/site-packages/pandas/core/internals/managers.py\u001b[0m in \u001b[0;36mas_array\u001b[0;34m(self, transpose, dtype, copy, na_value)\u001b[0m\n\u001b[1;32m    870\u001b[0m                     \u001b[0marr\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0marr\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mastype\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdtype\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcopy\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mFalse\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    871\u001b[0m         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 872\u001b[0;31m             \u001b[0marr\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_interleave\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdtype\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mdtype\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mna_value\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mna_value\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    873\u001b[0m             \u001b[0;31m# The underlying data was copied within _interleave\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    874\u001b[0m             \u001b[0mcopy\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mFalse\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/blaze/prom317/conda/envs/lab/lib/python3.7/site-packages/pandas/core/internals/managers.py\u001b[0m in \u001b[0;36m_interleave\u001b[0;34m(self, dtype, na_value)\u001b[0m\n\u001b[1;32m    909\u001b[0m                 \u001b[0marr\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mblk\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mvalues\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mto_numpy\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdtype\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mdtype\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mna_value\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mna_value\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    910\u001b[0m             \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 911\u001b[0;31m                 \u001b[0marr\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mblk\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_values\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdtype\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    912\u001b[0m             \u001b[0mresult\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mrl\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mindexer\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0marr\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    913\u001b[0m             \u001b[0mitemmask\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mrl\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mindexer\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;36m1\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/blaze/prom317/conda/envs/lab/lib/python3.7/site-packages/pandas/core/internals/blocks.py\u001b[0m in \u001b[0;36mget_values\u001b[0;34m(self, dtype)\u001b[0m\n\u001b[1;32m    249\u001b[0m         \"\"\"\n\u001b[1;32m    250\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mis_object_dtype\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdtype\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 251\u001b[0;31m             \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mvalues\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mastype\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mobject\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    252\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mvalues\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    253\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "for idx1, row1 in samData.iterrows():\n",
    "    count +=1\n",
    "    if count % 1000 == 0:\n",
    "        print(count)\n",
    "    try:\n",
    "        row2 = codingDF.loc[row1[15]]\n",
    "        pos = samData.loc[loc, 'pos']\n",
    "        samData.loc[idx, 'start_pos'] = pos - row2[5] \n",
    "        samData.loc[idx, 'end_pos'] = pos - row2[6]\n",
    "    except KeyError:\n",
    "        continue\n",
    "samData.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "hazardous-square",
   "metadata": {},
   "outputs": [],
   "source": [
    "samData = samData.loc[samData.start_pos.notnull()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "mounted-wilson",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>transcript_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSMUST00000193812.2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>3143476.0</td>\n",
       "      <td>3144545.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000102693.2\"; transcript_id ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUST00000082908.3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>transcript</td>\n",
       "      <td>3172239.0</td>\n",
       "      <td>3172348.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000064842.3\"; transcript_id ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUST00000162897.2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>3276124.0</td>\n",
       "      <td>3286567.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000051951.6\"; transcript_id ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUST00000159265.2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>3276746.0</td>\n",
       "      <td>3285855.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000051951.6\"; transcript_id ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUST00000070533.5</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>3284705.0</td>\n",
       "      <td>3741721.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000051951.6\"; transcript_id ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                         0        1           2          3          4  5  6  \\\n",
       "transcript_id                                                                 \n",
       "ENSMUST00000193812.2  chr1   HAVANA  transcript  3143476.0  3144545.0  .  +   \n",
       "ENSMUST00000082908.3  chr1  ENSEMBL  transcript  3172239.0  3172348.0  .  +   \n",
       "ENSMUST00000162897.2  chr1   HAVANA  transcript  3276124.0  3286567.0  .  -   \n",
       "ENSMUST00000159265.2  chr1   HAVANA  transcript  3276746.0  3285855.0  .  -   \n",
       "ENSMUST00000070533.5  chr1   HAVANA  transcript  3284705.0  3741721.0  .  -   \n",
       "\n",
       "                      7                                                  8  \n",
       "transcript_id                                                               \n",
       "ENSMUST00000193812.2  .  gene_id \"ENSMUSG00000102693.2\"; transcript_id ...  \n",
       "ENSMUST00000082908.3  .  gene_id \"ENSMUSG00000064842.3\"; transcript_id ...  \n",
       "ENSMUST00000162897.2  .  gene_id \"ENSMUSG00000051951.6\"; transcript_id ...  \n",
       "ENSMUST00000159265.2  .  gene_id \"ENSMUSG00000051951.6\"; transcript_id ...  \n",
       "ENSMUST00000070533.5  .  gene_id \"ENSMUSG00000051951.6\"; transcript_id ...  "
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gtfTranscriptData[\"transcript_id\"] = gtfTranscriptData.iloc[:,8].map(lambda p: p.split(';')[1].split(' ')[2][1:-1])\n",
    "gtfTranscriptData = gtfTranscriptData.set_index('transcript_id')\n",
    "gtfTranscriptData.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "freelance-omaha",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0\n",
      "10000\n",
      "20000\n",
      "30000\n",
      "40000\n",
      "50000\n",
      "60000\n",
      "70000\n",
      "80000\n",
      "90000\n",
      "100000\n",
      "110000\n",
      "120000\n",
      "130000\n",
      "140000\n",
      "150000\n",
      "160000\n",
      "170000\n",
      "180000\n",
      "190000\n",
      "200000\n",
      "210000\n",
      "220000\n",
      "230000\n",
      "240000\n",
      "250000\n",
      "260000\n",
      "270000\n",
      "280000\n",
      "290000\n",
      "300000\n",
      "310000\n",
      "320000\n",
      "330000\n",
      "340000\n",
      "350000\n",
      "360000\n",
      "370000\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>10</th>\n",
       "      <th>11</th>\n",
       "      <th>12</th>\n",
       "      <th>13</th>\n",
       "      <th>14</th>\n",
       "      <th>distance</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SRR458756.2.1104</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSMUST00000129380.2</td>\n",
       "      <td>117</td>\n",
       "      <td>1</td>\n",
       "      <td>30M</td>\n",
       "      <td>*</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>CGAGTGGAGACTGGTGTTCTCAAGCCTGGC</td>\n",
       "      <td>IIIIIIIIIIIIIIIIIIGIIIIIIIIIIH</td>\n",
       "      <td>NH:i:4</td>\n",
       "      <td>HI:i:1</td>\n",
       "      <td>AS:i:29</td>\n",
       "      <td>nM:i:0</td>\n",
       "      <td>1052.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SRR458756.2.1104</td>\n",
       "      <td>256</td>\n",
       "      <td>ENSMUST00000042235.15</td>\n",
       "      <td>916</td>\n",
       "      <td>1</td>\n",
       "      <td>30M</td>\n",
       "      <td>*</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>CGAGTGGAGACTGGTGTTCTCAAGCCTGGC</td>\n",
       "      <td>IIIIIIIIIIIIIIIIIIGIIIIIIIIIIH</td>\n",
       "      <td>NH:i:4</td>\n",
       "      <td>HI:i:2</td>\n",
       "      <td>AS:i:29</td>\n",
       "      <td>nM:i:0</td>\n",
       "      <td>-157.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SRR458756.2.1104</td>\n",
       "      <td>256</td>\n",
       "      <td>ENSMUST00000144523.8</td>\n",
       "      <td>919</td>\n",
       "      <td>1</td>\n",
       "      <td>30M</td>\n",
       "      <td>*</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>CGAGTGGAGACTGGTGTTCTCAAGCCTGGC</td>\n",
       "      <td>IIIIIIIIIIIIIIIIIIGIIIIIIIIIIH</td>\n",
       "      <td>NH:i:4</td>\n",
       "      <td>HI:i:3</td>\n",
       "      <td>AS:i:29</td>\n",
       "      <td>nM:i:0</td>\n",
       "      <td>882.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SRR458756.2.1104</td>\n",
       "      <td>256</td>\n",
       "      <td>ENSMUST00000150960.2</td>\n",
       "      <td>823</td>\n",
       "      <td>1</td>\n",
       "      <td>30M</td>\n",
       "      <td>*</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>CGAGTGGAGACTGGTGTTCTCAAGCCTGGC</td>\n",
       "      <td>IIIIIIIIIIIIIIIIIIGIIIIIIIIIIH</td>\n",
       "      <td>NH:i:4</td>\n",
       "      <td>HI:i:4</td>\n",
       "      <td>AS:i:29</td>\n",
       "      <td>nM:i:0</td>\n",
       "      <td>975.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SRR458756.2.1167</td>\n",
       "      <td>256</td>\n",
       "      <td>ENSMUST00000042235.15</td>\n",
       "      <td>1390</td>\n",
       "      <td>0</td>\n",
       "      <td>32M</td>\n",
       "      <td>*</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>TTTGCTGTTCGTGACATGAGGCAGACAGTTGC</td>\n",
       "      <td>IIIIIIIIIIIIIIIIIIHIIIIDIIIBIIHG</td>\n",
       "      <td>NH:i:7</td>\n",
       "      <td>HI:i:3</td>\n",
       "      <td>AS:i:31</td>\n",
       "      <td>nM:i:0</td>\n",
       "      <td>317.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  0    1                      2     3  4    5  6  7  8  \\\n",
       "0  SRR458756.2.1104    0   ENSMUST00000129380.2   117  1  30M  *  0  0   \n",
       "1  SRR458756.2.1104  256  ENSMUST00000042235.15   916  1  30M  *  0  0   \n",
       "2  SRR458756.2.1104  256   ENSMUST00000144523.8   919  1  30M  *  0  0   \n",
       "3  SRR458756.2.1104  256   ENSMUST00000150960.2   823  1  30M  *  0  0   \n",
       "4  SRR458756.2.1167  256  ENSMUST00000042235.15  1390  0  32M  *  0  0   \n",
       "\n",
       "                                  9                                10      11  \\\n",
       "0    CGAGTGGAGACTGGTGTTCTCAAGCCTGGC    IIIIIIIIIIIIIIIIIIGIIIIIIIIIIH  NH:i:4   \n",
       "1    CGAGTGGAGACTGGTGTTCTCAAGCCTGGC    IIIIIIIIIIIIIIIIIIGIIIIIIIIIIH  NH:i:4   \n",
       "2    CGAGTGGAGACTGGTGTTCTCAAGCCTGGC    IIIIIIIIIIIIIIIIIIGIIIIIIIIIIH  NH:i:4   \n",
       "3    CGAGTGGAGACTGGTGTTCTCAAGCCTGGC    IIIIIIIIIIIIIIIIIIGIIIIIIIIIIH  NH:i:4   \n",
       "4  TTTGCTGTTCGTGACATGAGGCAGACAGTTGC  IIIIIIIIIIIIIIIIIIHIIIIDIIIBIIHG  NH:i:7   \n",
       "\n",
       "       12       13      14  distance  \n",
       "0  HI:i:1  AS:i:29  nM:i:0    1052.0  \n",
       "1  HI:i:2  AS:i:29  nM:i:0    -157.0  \n",
       "2  HI:i:3  AS:i:29  nM:i:0     882.0  \n",
       "3  HI:i:4  AS:i:29  nM:i:0     975.0  \n",
       "4  HI:i:3  AS:i:31  nM:i:0     317.0  "
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "count = 0\n",
    "samData['distance'] = np.nan\n",
    "for idx1, row1 in samData.iterrows():\n",
    "    if count % 10000 == 0:\n",
    "        print(count)\n",
    "    try:\n",
    "        row2 = gtfTranscriptData.loc[row1[2]]\n",
    "        samData.loc[idx1,'distance'] = 78387934 - (row2[4]-row1[3]+1)\n",
    "        count = count +1\n",
    "    except KeyError:\n",
    "        if count %10000 == 0:\n",
    "            print('Error')\n",
    "        continue\n",
    "samData.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "affected-medium",
   "metadata": {},
   "outputs": [],
   "source": [
    "samData_drop = samData.loc[samData.distance.notnull()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "extended-grammar",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>10</th>\n",
       "      <th>11</th>\n",
       "      <th>12</th>\n",
       "      <th>13</th>\n",
       "      <th>14</th>\n",
       "      <th>distance</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SRR458756.2.1104</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSMUST00000129380.2</td>\n",
       "      <td>117</td>\n",
       "      <td>1</td>\n",
       "      <td>30M</td>\n",
       "      <td>*</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>CGAGTGGAGACTGGTGTTCTCAAGCCTGGC</td>\n",
       "      <td>IIIIIIIIIIIIIIIIIIGIIIIIIIIIIH</td>\n",
       "      <td>NH:i:4</td>\n",
       "      <td>HI:i:1</td>\n",
       "      <td>AS:i:29</td>\n",
       "      <td>nM:i:0</td>\n",
       "      <td>1052.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SRR458756.2.1104</td>\n",
       "      <td>256</td>\n",
       "      <td>ENSMUST00000042235.15</td>\n",
       "      <td>916</td>\n",
       "      <td>1</td>\n",
       "      <td>30M</td>\n",
       "      <td>*</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>CGAGTGGAGACTGGTGTTCTCAAGCCTGGC</td>\n",
       "      <td>IIIIIIIIIIIIIIIIIIGIIIIIIIIIIH</td>\n",
       "      <td>NH:i:4</td>\n",
       "      <td>HI:i:2</td>\n",
       "      <td>AS:i:29</td>\n",
       "      <td>nM:i:0</td>\n",
       "      <td>-157.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SRR458756.2.1104</td>\n",
       "      <td>256</td>\n",
       "      <td>ENSMUST00000144523.8</td>\n",
       "      <td>919</td>\n",
       "      <td>1</td>\n",
       "      <td>30M</td>\n",
       "      <td>*</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>CGAGTGGAGACTGGTGTTCTCAAGCCTGGC</td>\n",
       "      <td>IIIIIIIIIIIIIIIIIIGIIIIIIIIIIH</td>\n",
       "      <td>NH:i:4</td>\n",
       "      <td>HI:i:3</td>\n",
       "      <td>AS:i:29</td>\n",
       "      <td>nM:i:0</td>\n",
       "      <td>882.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SRR458756.2.1104</td>\n",
       "      <td>256</td>\n",
       "      <td>ENSMUST00000150960.2</td>\n",
       "      <td>823</td>\n",
       "      <td>1</td>\n",
       "      <td>30M</td>\n",
       "      <td>*</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>CGAGTGGAGACTGGTGTTCTCAAGCCTGGC</td>\n",
       "      <td>IIIIIIIIIIIIIIIIIIGIIIIIIIIIIH</td>\n",
       "      <td>NH:i:4</td>\n",
       "      <td>HI:i:4</td>\n",
       "      <td>AS:i:29</td>\n",
       "      <td>nM:i:0</td>\n",
       "      <td>975.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SRR458756.2.1167</td>\n",
       "      <td>256</td>\n",
       "      <td>ENSMUST00000042235.15</td>\n",
       "      <td>1390</td>\n",
       "      <td>0</td>\n",
       "      <td>32M</td>\n",
       "      <td>*</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>TTTGCTGTTCGTGACATGAGGCAGACAGTTGC</td>\n",
       "      <td>IIIIIIIIIIIIIIIIIIHIIIIDIIIBIIHG</td>\n",
       "      <td>NH:i:7</td>\n",
       "      <td>HI:i:3</td>\n",
       "      <td>AS:i:31</td>\n",
       "      <td>nM:i:0</td>\n",
       "      <td>317.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  0    1                      2     3  4    5  6  7  8  \\\n",
       "0  SRR458756.2.1104    0   ENSMUST00000129380.2   117  1  30M  *  0  0   \n",
       "1  SRR458756.2.1104  256  ENSMUST00000042235.15   916  1  30M  *  0  0   \n",
       "2  SRR458756.2.1104  256   ENSMUST00000144523.8   919  1  30M  *  0  0   \n",
       "3  SRR458756.2.1104  256   ENSMUST00000150960.2   823  1  30M  *  0  0   \n",
       "4  SRR458756.2.1167  256  ENSMUST00000042235.15  1390  0  32M  *  0  0   \n",
       "\n",
       "                                  9                                10      11  \\\n",
       "0    CGAGTGGAGACTGGTGTTCTCAAGCCTGGC    IIIIIIIIIIIIIIIIIIGIIIIIIIIIIH  NH:i:4   \n",
       "1    CGAGTGGAGACTGGTGTTCTCAAGCCTGGC    IIIIIIIIIIIIIIIIIIGIIIIIIIIIIH  NH:i:4   \n",
       "2    CGAGTGGAGACTGGTGTTCTCAAGCCTGGC    IIIIIIIIIIIIIIIIIIGIIIIIIIIIIH  NH:i:4   \n",
       "3    CGAGTGGAGACTGGTGTTCTCAAGCCTGGC    IIIIIIIIIIIIIIIIIIGIIIIIIIIIIH  NH:i:4   \n",
       "4  TTTGCTGTTCGTGACATGAGGCAGACAGTTGC  IIIIIIIIIIIIIIIIIIHIIIIDIIIBIIHG  NH:i:7   \n",
       "\n",
       "       12       13      14  distance  \n",
       "0  HI:i:1  AS:i:29  nM:i:0    1052.0  \n",
       "1  HI:i:2  AS:i:29  nM:i:0    -157.0  \n",
       "2  HI:i:3  AS:i:29  nM:i:0     882.0  \n",
       "3  HI:i:4  AS:i:29  nM:i:0     975.0  \n",
       "4  HI:i:3  AS:i:31  nM:i:0     317.0  "
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "samData_drop.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "prerequisite-thumb",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
