{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "august-paintball",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>qname</th>\n",
       "      <th>flag</th>\n",
       "      <th>rname</th>\n",
       "      <th>pos</th>\n",
       "      <th>mapq</th>\n",
       "      <th>cigar</th>\n",
       "      <th>rnext</th>\n",
       "      <th>pnext</th>\n",
       "      <th>tlen</th>\n",
       "      <th>seq</th>\n",
       "      <th>qual</th>\n",
       "      <th>opt1</th>\n",
       "      <th>opt2</th>\n",
       "      <th>opt3</th>\n",
       "      <th>opt4</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SRR458756.2.803</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSMUST00000240377.1|ENSMUSG00000119584.1|OTTM...</td>\n",
       "      <td>295</td>\n",
       "      <td>255</td>\n",
       "      <td>15M</td>\n",
       "      <td>*</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>TGGTGACTCTAGATA</td>\n",
       "      <td>B@@@@AAAAAAAAAA</td>\n",
       "      <td>NH:i:1</td>\n",
       "      <td>HI:i:1</td>\n",
       "      <td>AS:i:14</td>\n",
       "      <td>nM:i:0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SRR458756.2.918</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSMUST00000091752.5|ENSMUSG00000069310.5|OTTM...</td>\n",
       "      <td>100</td>\n",
       "      <td>255</td>\n",
       "      <td>30M</td>\n",
       "      <td>*</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>CCAAGAATGGCTGGTACTAAGCAGACCGCT</td>\n",
       "      <td>A@A@@@A@@@?*.56A&lt;2A&lt;+;?*A??*?8</td>\n",
       "      <td>NH:i:1</td>\n",
       "      <td>HI:i:1</td>\n",
       "      <td>AS:i:27</td>\n",
       "      <td>nM:i:1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SRR458756.2.919</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSMUST00000240377.1|ENSMUSG00000119584.1|OTTM...</td>\n",
       "      <td>815</td>\n",
       "      <td>255</td>\n",
       "      <td>28M</td>\n",
       "      <td>*</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>AAAAAATTAGAGTGTTCAAAGCAGGCCC</td>\n",
       "      <td>IIIIIIIHIHGGBG8AFFFFDGGBBGGG</td>\n",
       "      <td>NH:i:1</td>\n",
       "      <td>HI:i:1</td>\n",
       "      <td>AS:i:27</td>\n",
       "      <td>nM:i:0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SRR458756.2.922</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSMUST00000240377.1|ENSMUSG00000119584.1|OTTM...</td>\n",
       "      <td>292</td>\n",
       "      <td>255</td>\n",
       "      <td>30M</td>\n",
       "      <td>*</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>CTTTGGCGACTCTAGATAACCTCGGGCCGA</td>\n",
       "      <td>##############################</td>\n",
       "      <td>NH:i:1</td>\n",
       "      <td>HI:i:1</td>\n",
       "      <td>AS:i:27</td>\n",
       "      <td>nM:i:1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SRR458756.2.923</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSMUST00000023741.16|ENSMUSG00000048154.18|OT...</td>\n",
       "      <td>4117</td>\n",
       "      <td>3</td>\n",
       "      <td>1S17M</td>\n",
       "      <td>*</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>AGTGGCTGATATTGATAG</td>\n",
       "      <td>FC?BCGGGBBGEG@GDG&gt;</td>\n",
       "      <td>NH:i:2</td>\n",
       "      <td>HI:i:1</td>\n",
       "      <td>AS:i:16</td>\n",
       "      <td>nM:i:0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             qname  flag                                              rname  \\\n",
       "0  SRR458756.2.803     0  ENSMUST00000240377.1|ENSMUSG00000119584.1|OTTM...   \n",
       "1  SRR458756.2.918     0  ENSMUST00000091752.5|ENSMUSG00000069310.5|OTTM...   \n",
       "2  SRR458756.2.919     0  ENSMUST00000240377.1|ENSMUSG00000119584.1|OTTM...   \n",
       "3  SRR458756.2.922     0  ENSMUST00000240377.1|ENSMUSG00000119584.1|OTTM...   \n",
       "4  SRR458756.2.923     0  ENSMUST00000023741.16|ENSMUSG00000048154.18|OT...   \n",
       "\n",
       "    pos  mapq  cigar rnext  pnext  tlen                             seq  \\\n",
       "0   295   255    15M     *      0     0                 TGGTGACTCTAGATA   \n",
       "1   100   255    30M     *      0     0  CCAAGAATGGCTGGTACTAAGCAGACCGCT   \n",
       "2   815   255    28M     *      0     0    AAAAAATTAGAGTGTTCAAAGCAGGCCC   \n",
       "3   292   255    30M     *      0     0  CTTTGGCGACTCTAGATAACCTCGGGCCGA   \n",
       "4  4117     3  1S17M     *      0     0              AGTGGCTGATATTGATAG   \n",
       "\n",
       "                             qual    opt1    opt2     opt3    opt4  \n",
       "0                 B@@@@AAAAAAAAAA  NH:i:1  HI:i:1  AS:i:14  nM:i:0  \n",
       "1  A@A@@@A@@@?*.56A<2A<+;?*A??*?8  NH:i:1  HI:i:1  AS:i:27  nM:i:1  \n",
       "2    IIIIIIIHIHGGBG8AFFFFDGGBBGGG  NH:i:1  HI:i:1  AS:i:27  nM:i:0  \n",
       "3  ##############################  NH:i:1  HI:i:1  AS:i:27  nM:i:1  \n",
       "4              FC?BCGGGBBGEG@GDG>  NH:i:2  HI:i:1  AS:i:16  nM:i:0  "
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import numpy as np\n",
    "import pandas as pd \n",
    "\n",
    "samData = pd.read_csv(\"/qbio/nest/prom317/tutorial/transcriptomeAlignment/RPF_siLuc/ForwardStrand.out.sam\", sep = '\\t', names=['qname', 'flag', 'rname', 'pos', 'mapq', 'cigar', 'rnext', 'pnext', 'tlen', 'seq', 'qual','opt1', 'opt2', 'opt3', 'opt4'])\n",
    "samData.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "removed-michael",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(43076098, 17)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>qname</th>\n",
       "      <th>flag</th>\n",
       "      <th>rname</th>\n",
       "      <th>pos</th>\n",
       "      <th>mapq</th>\n",
       "      <th>cigar</th>\n",
       "      <th>rnext</th>\n",
       "      <th>pnext</th>\n",
       "      <th>tlen</th>\n",
       "      <th>seq</th>\n",
       "      <th>qual</th>\n",
       "      <th>opt1</th>\n",
       "      <th>opt2</th>\n",
       "      <th>opt3</th>\n",
       "      <th>opt4</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SRR458756.2.803</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSMUST00000240377.1|ENSMUSG00000119584.1|OTTM...</td>\n",
       "      <td>295</td>\n",
       "      <td>255</td>\n",
       "      <td>15M</td>\n",
       "      <td>*</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>TGGTGACTCTAGATA</td>\n",
       "      <td>B@@@@AAAAAAAAAA</td>\n",
       "      <td>NH:i:1</td>\n",
       "      <td>HI:i:1</td>\n",
       "      <td>AS:i:14</td>\n",
       "      <td>nM:i:0</td>\n",
       "      <td>ENSMUST00000240377.1</td>\n",
       "      <td>ENSMUSG00000119584.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SRR458756.2.918</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSMUST00000091752.5|ENSMUSG00000069310.5|OTTM...</td>\n",
       "      <td>100</td>\n",
       "      <td>255</td>\n",
       "      <td>30M</td>\n",
       "      <td>*</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>CCAAGAATGGCTGGTACTAAGCAGACCGCT</td>\n",
       "      <td>A@A@@@A@@@?*.56A&lt;2A&lt;+;?*A??*?8</td>\n",
       "      <td>NH:i:1</td>\n",
       "      <td>HI:i:1</td>\n",
       "      <td>AS:i:27</td>\n",
       "      <td>nM:i:1</td>\n",
       "      <td>ENSMUST00000091752.5</td>\n",
       "      <td>ENSMUSG00000069310.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SRR458756.2.919</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSMUST00000240377.1|ENSMUSG00000119584.1|OTTM...</td>\n",
       "      <td>815</td>\n",
       "      <td>255</td>\n",
       "      <td>28M</td>\n",
       "      <td>*</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>AAAAAATTAGAGTGTTCAAAGCAGGCCC</td>\n",
       "      <td>IIIIIIIHIHGGBG8AFFFFDGGBBGGG</td>\n",
       "      <td>NH:i:1</td>\n",
       "      <td>HI:i:1</td>\n",
       "      <td>AS:i:27</td>\n",
       "      <td>nM:i:0</td>\n",
       "      <td>ENSMUST00000240377.1</td>\n",
       "      <td>ENSMUSG00000119584.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SRR458756.2.922</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSMUST00000240377.1|ENSMUSG00000119584.1|OTTM...</td>\n",
       "      <td>292</td>\n",
       "      <td>255</td>\n",
       "      <td>30M</td>\n",
       "      <td>*</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>CTTTGGCGACTCTAGATAACCTCGGGCCGA</td>\n",
       "      <td>##############################</td>\n",
       "      <td>NH:i:1</td>\n",
       "      <td>HI:i:1</td>\n",
       "      <td>AS:i:27</td>\n",
       "      <td>nM:i:1</td>\n",
       "      <td>ENSMUST00000240377.1</td>\n",
       "      <td>ENSMUSG00000119584.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SRR458756.2.923</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSMUST00000023741.16|ENSMUSG00000048154.18|OT...</td>\n",
       "      <td>4117</td>\n",
       "      <td>3</td>\n",
       "      <td>1S17M</td>\n",
       "      <td>*</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>AGTGGCTGATATTGATAG</td>\n",
       "      <td>FC?BCGGGBBGEG@GDG&gt;</td>\n",
       "      <td>NH:i:2</td>\n",
       "      <td>HI:i:1</td>\n",
       "      <td>AS:i:16</td>\n",
       "      <td>nM:i:0</td>\n",
       "      <td>ENSMUST00000023741.16</td>\n",
       "      <td>ENSMUSG00000048154.18</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             qname  flag                                              rname  \\\n",
       "0  SRR458756.2.803     0  ENSMUST00000240377.1|ENSMUSG00000119584.1|OTTM...   \n",
       "1  SRR458756.2.918     0  ENSMUST00000091752.5|ENSMUSG00000069310.5|OTTM...   \n",
       "2  SRR458756.2.919     0  ENSMUST00000240377.1|ENSMUSG00000119584.1|OTTM...   \n",
       "3  SRR458756.2.922     0  ENSMUST00000240377.1|ENSMUSG00000119584.1|OTTM...   \n",
       "4  SRR458756.2.923     0  ENSMUST00000023741.16|ENSMUSG00000048154.18|OT...   \n",
       "\n",
       "    pos  mapq  cigar rnext  pnext  tlen                             seq  \\\n",
       "0   295   255    15M     *      0     0                 TGGTGACTCTAGATA   \n",
       "1   100   255    30M     *      0     0  CCAAGAATGGCTGGTACTAAGCAGACCGCT   \n",
       "2   815   255    28M     *      0     0    AAAAAATTAGAGTGTTCAAAGCAGGCCC   \n",
       "3   292   255    30M     *      0     0  CTTTGGCGACTCTAGATAACCTCGGGCCGA   \n",
       "4  4117     3  1S17M     *      0     0              AGTGGCTGATATTGATAG   \n",
       "\n",
       "                             qual    opt1    opt2     opt3    opt4  \\\n",
       "0                 B@@@@AAAAAAAAAA  NH:i:1  HI:i:1  AS:i:14  nM:i:0   \n",
       "1  A@A@@@A@@@?*.56A<2A<+;?*A??*?8  NH:i:1  HI:i:1  AS:i:27  nM:i:1   \n",
       "2    IIIIIIIHIHGGBG8AFFFFDGGBBGGG  NH:i:1  HI:i:1  AS:i:27  nM:i:0   \n",
       "3  ##############################  NH:i:1  HI:i:1  AS:i:27  nM:i:1   \n",
       "4              FC?BCGGGBBGEG@GDG>  NH:i:2  HI:i:1  AS:i:16  nM:i:0   \n",
       "\n",
       "           transcript_id                gene_id  \n",
       "0   ENSMUST00000240377.1   ENSMUSG00000119584.1  \n",
       "1   ENSMUST00000091752.5   ENSMUSG00000069310.5  \n",
       "2   ENSMUST00000240377.1   ENSMUSG00000119584.1  \n",
       "3   ENSMUST00000240377.1   ENSMUSG00000119584.1  \n",
       "4  ENSMUST00000023741.16  ENSMUSG00000048154.18  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "samData['transcript_id'] = samData.rname.map(lambda p : p.split('|')[0])\n",
    "samData['gene_id'] = samData.rname.map(lambda p : p.split('|')[1])\n",
    "print(samData.shape)\n",
    "samData.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "contrary-value",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seq_name</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attribute</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>3143476.0</td>\n",
       "      <td>3144545.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000102693.2\"; gene_type \"TEC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>3143476.0</td>\n",
       "      <td>3144545.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000102693.2\"; transcript_id ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>3143476.0</td>\n",
       "      <td>3144545.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000102693.2\"; transcript_id ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>gene</td>\n",
       "      <td>3172239.0</td>\n",
       "      <td>3172348.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000064842.3\"; gene_type \"snR...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>transcript</td>\n",
       "      <td>3172239.0</td>\n",
       "      <td>3172348.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000064842.3\"; transcript_id ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  seq_name   source     feature      start        end score strand frame  \\\n",
       "5     chr1   HAVANA        gene  3143476.0  3144545.0     .      +     .   \n",
       "6     chr1   HAVANA  transcript  3143476.0  3144545.0     .      +     .   \n",
       "7     chr1   HAVANA        exon  3143476.0  3144545.0     .      +     .   \n",
       "8     chr1  ENSEMBL        gene  3172239.0  3172348.0     .      +     .   \n",
       "9     chr1  ENSEMBL  transcript  3172239.0  3172348.0     .      +     .   \n",
       "\n",
       "                                           attribute  \n",
       "5  gene_id \"ENSMUSG00000102693.2\"; gene_type \"TEC...  \n",
       "6  gene_id \"ENSMUSG00000102693.2\"; transcript_id ...  \n",
       "7  gene_id \"ENSMUSG00000102693.2\"; transcript_id ...  \n",
       "8  gene_id \"ENSMUSG00000064842.3\"; gene_type \"snR...  \n",
       "9  gene_id \"ENSMUSG00000064842.3\"; transcript_id ...  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gtfData = pd.read_table(\"/qbio/nest/prom317/tutorial/references/GRCm39.genome.gtf\", sep = '\\t', names = [\"seq_name\",\"source\",\"feature\",\"start\",\"end\",\"score\",\"strand\",\"frame\",\"attribute\"])\n",
    "gtfData.iloc[5:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "center-juvenile",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>qname</th>\n",
       "      <th>flag</th>\n",
       "      <th>rname</th>\n",
       "      <th>pos</th>\n",
       "      <th>mapq</th>\n",
       "      <th>cigar</th>\n",
       "      <th>rnext</th>\n",
       "      <th>pnext</th>\n",
       "      <th>tlen</th>\n",
       "      <th>seq</th>\n",
       "      <th>qual</th>\n",
       "      <th>opt1</th>\n",
       "      <th>opt2</th>\n",
       "      <th>opt3</th>\n",
       "      <th>opt4</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>start_pos</th>\n",
       "      <th>end_pos</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SRR458756.2.803</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSMUST00000240377.1|ENSMUSG00000119584.1|OTTM...</td>\n",
       "      <td>295</td>\n",
       "      <td>255</td>\n",
       "      <td>15M</td>\n",
       "      <td>*</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>TGGTGACTCTAGATA</td>\n",
       "      <td>B@@@@AAAAAAAAAA</td>\n",
       "      <td>NH:i:1</td>\n",
       "      <td>HI:i:1</td>\n",
       "      <td>AS:i:14</td>\n",
       "      <td>nM:i:0</td>\n",
       "      <td>ENSMUST00000240377.1</td>\n",
       "      <td>ENSMUSG00000119584.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SRR458756.2.918</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSMUST00000091752.5|ENSMUSG00000069310.5|OTTM...</td>\n",
       "      <td>100</td>\n",
       "      <td>255</td>\n",
       "      <td>30M</td>\n",
       "      <td>*</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>CCAAGAATGGCTGGTACTAAGCAGACCGCT</td>\n",
       "      <td>A@A@@@A@@@?*.56A&lt;2A&lt;+;?*A??*?8</td>\n",
       "      <td>NH:i:1</td>\n",
       "      <td>HI:i:1</td>\n",
       "      <td>AS:i:27</td>\n",
       "      <td>nM:i:1</td>\n",
       "      <td>ENSMUST00000091752.5</td>\n",
       "      <td>ENSMUSG00000069310.5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SRR458756.2.919</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSMUST00000240377.1|ENSMUSG00000119584.1|OTTM...</td>\n",
       "      <td>815</td>\n",
       "      <td>255</td>\n",
       "      <td>28M</td>\n",
       "      <td>*</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>AAAAAATTAGAGTGTTCAAAGCAGGCCC</td>\n",
       "      <td>IIIIIIIHIHGGBG8AFFFFDGGBBGGG</td>\n",
       "      <td>NH:i:1</td>\n",
       "      <td>HI:i:1</td>\n",
       "      <td>AS:i:27</td>\n",
       "      <td>nM:i:0</td>\n",
       "      <td>ENSMUST00000240377.1</td>\n",
       "      <td>ENSMUSG00000119584.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SRR458756.2.922</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSMUST00000240377.1|ENSMUSG00000119584.1|OTTM...</td>\n",
       "      <td>292</td>\n",
       "      <td>255</td>\n",
       "      <td>30M</td>\n",
       "      <td>*</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>CTTTGGCGACTCTAGATAACCTCGGGCCGA</td>\n",
       "      <td>##############################</td>\n",
       "      <td>NH:i:1</td>\n",
       "      <td>HI:i:1</td>\n",
       "      <td>AS:i:27</td>\n",
       "      <td>nM:i:1</td>\n",
       "      <td>ENSMUST00000240377.1</td>\n",
       "      <td>ENSMUSG00000119584.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SRR458756.2.923</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSMUST00000023741.16|ENSMUSG00000048154.18|OT...</td>\n",
       "      <td>4117</td>\n",
       "      <td>3</td>\n",
       "      <td>1S17M</td>\n",
       "      <td>*</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>AGTGGCTGATATTGATAG</td>\n",
       "      <td>FC?BCGGGBBGEG@GDG&gt;</td>\n",
       "      <td>NH:i:2</td>\n",
       "      <td>HI:i:1</td>\n",
       "      <td>AS:i:16</td>\n",
       "      <td>nM:i:0</td>\n",
       "      <td>ENSMUST00000023741.16</td>\n",
       "      <td>ENSMUSG00000048154.18</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             qname  flag                                              rname  \\\n",
       "0  SRR458756.2.803     0  ENSMUST00000240377.1|ENSMUSG00000119584.1|OTTM...   \n",
       "1  SRR458756.2.918     0  ENSMUST00000091752.5|ENSMUSG00000069310.5|OTTM...   \n",
       "2  SRR458756.2.919     0  ENSMUST00000240377.1|ENSMUSG00000119584.1|OTTM...   \n",
       "3  SRR458756.2.922     0  ENSMUST00000240377.1|ENSMUSG00000119584.1|OTTM...   \n",
       "4  SRR458756.2.923     0  ENSMUST00000023741.16|ENSMUSG00000048154.18|OT...   \n",
       "\n",
       "    pos  mapq  cigar rnext  pnext  tlen                             seq  \\\n",
       "0   295   255    15M     *      0     0                 TGGTGACTCTAGATA   \n",
       "1   100   255    30M     *      0     0  CCAAGAATGGCTGGTACTAAGCAGACCGCT   \n",
       "2   815   255    28M     *      0     0    AAAAAATTAGAGTGTTCAAAGCAGGCCC   \n",
       "3   292   255    30M     *      0     0  CTTTGGCGACTCTAGATAACCTCGGGCCGA   \n",
       "4  4117     3  1S17M     *      0     0              AGTGGCTGATATTGATAG   \n",
       "\n",
       "                             qual    opt1    opt2     opt3    opt4  \\\n",
       "0                 B@@@@AAAAAAAAAA  NH:i:1  HI:i:1  AS:i:14  nM:i:0   \n",
       "1  A@A@@@A@@@?*.56A<2A<+;?*A??*?8  NH:i:1  HI:i:1  AS:i:27  nM:i:1   \n",
       "2    IIIIIIIHIHGGBG8AFFFFDGGBBGGG  NH:i:1  HI:i:1  AS:i:27  nM:i:0   \n",
       "3  ##############################  NH:i:1  HI:i:1  AS:i:27  nM:i:1   \n",
       "4              FC?BCGGGBBGEG@GDG>  NH:i:2  HI:i:1  AS:i:16  nM:i:0   \n",
       "\n",
       "           transcript_id                gene_id  start_pos  end_pos  \n",
       "0   ENSMUST00000240377.1   ENSMUSG00000119584.1        NaN      NaN  \n",
       "1   ENSMUST00000091752.5   ENSMUSG00000069310.5        NaN      NaN  \n",
       "2   ENSMUST00000240377.1   ENSMUSG00000119584.1        NaN      NaN  \n",
       "3   ENSMUST00000240377.1   ENSMUSG00000119584.1        NaN      NaN  \n",
       "4  ENSMUST00000023741.16  ENSMUSG00000048154.18        NaN      NaN  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "samData['start_pos'] = np.nan\n",
    "samData['end_pos'] = np.nan\n",
    "samData.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "equipped-berkeley",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "ENSMUSG00000052613.17    2003\n",
       "ENSMUSG00000069601.15    1772\n",
       "ENSMUSG00000021843.18    1695\n",
       "ENSMUSG00000064302.14    1411\n",
       "ENSMUSG00000033788.16    1411\n",
       "                         ... \n",
       "ENSMUSG00000119588.1        3\n",
       "ENSMUSG00000117051.2        3\n",
       "ENSMUSG00000107178.2        3\n",
       "ENSMUSG00000070610.5        3\n",
       "ENSMUSG00000117980.2        3\n",
       "Name: gene_id, Length: 26876, dtype: int64"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gtfPlusData = gtfData.loc[gtfData.strand == \"+\"]\n",
    "del gtfData\n",
    "gtfPlusData['gene_id'] = gtfPlusData.attribute.map(lambda p: p.split(';')[0].split(' ')[1][1:-1])\n",
    "gtfPlusData.gene_id.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "accessory-jaguar",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['ENSMUST00000240377.1' 'ENSMUST00000091752.5' 'ENSMUST00000023741.16' ...\n",
      " 'ENSMUST00000141667.2' 'ENSMUST00000099986.3' 'ENSMUST00000225350.2']\n"
     ]
    }
   ],
   "source": [
    "transcript_list = samData['transcript_id'].unique()\n",
    "transDict = {}\n",
    "for ts in transcript_list:\n",
    "    transDict[ts] = [[],[]]\n",
    "print(transcript_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "trained-essex",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/blaze/prom317/conda/envs/lab/lib/python3.7/site-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  \n",
      "/blaze/prom317/conda/envs/lab/lib/python3.7/site-packages/ipykernel_launcher.py:4: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  after removing the cwd from sys.path.\n",
      "/blaze/prom317/conda/envs/lab/lib/python3.7/site-packages/ipykernel_launcher.py:6: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  \n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seq_name</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attribute</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>361</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>4878137.0</td>\n",
       "      <td>4878139.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"ENSMUSG00000025903.15\"; transcript_id...</td>\n",
       "      <td>ENSMUSG00000025903.15</td>\n",
       "      <td>ENSMUST00000134384.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>386</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>4878137.0</td>\n",
       "      <td>4878139.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"ENSMUSG00000025903.15\"; transcript_id...</td>\n",
       "      <td>ENSMUSG00000025903.15</td>\n",
       "      <td>ENSMUST00000027036.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>409</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>4878137.0</td>\n",
       "      <td>4878139.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"ENSMUSG00000025903.15\"; transcript_id...</td>\n",
       "      <td>ENSMUSG00000025903.15</td>\n",
       "      <td>ENSMUST00000150971.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>430</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>4878137.0</td>\n",
       "      <td>4878139.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"ENSMUSG00000025903.15\"; transcript_id...</td>\n",
       "      <td>ENSMUSG00000025903.15</td>\n",
       "      <td>ENSMUST00000119612.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>449</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>4878137.0</td>\n",
       "      <td>4878139.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"ENSMUSG00000025903.15\"; transcript_id...</td>\n",
       "      <td>ENSMUSG00000025903.15</td>\n",
       "      <td>ENSMUST00000137887.8</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    seq_name  source      feature      start        end score strand frame  \\\n",
       "361     chr1  HAVANA  start_codon  4878137.0  4878139.0     .      +     0   \n",
       "386     chr1  HAVANA  start_codon  4878137.0  4878139.0     .      +     0   \n",
       "409     chr1  HAVANA  start_codon  4878137.0  4878139.0     .      +     0   \n",
       "430     chr1  HAVANA  start_codon  4878137.0  4878139.0     .      +     0   \n",
       "449     chr1  HAVANA  start_codon  4878137.0  4878139.0     .      +     0   \n",
       "\n",
       "                                             attribute                gene_id  \\\n",
       "361  gene_id \"ENSMUSG00000025903.15\"; transcript_id...  ENSMUSG00000025903.15   \n",
       "386  gene_id \"ENSMUSG00000025903.15\"; transcript_id...  ENSMUSG00000025903.15   \n",
       "409  gene_id \"ENSMUSG00000025903.15\"; transcript_id...  ENSMUSG00000025903.15   \n",
       "430  gene_id \"ENSMUSG00000025903.15\"; transcript_id...  ENSMUSG00000025903.15   \n",
       "449  gene_id \"ENSMUSG00000025903.15\"; transcript_id...  ENSMUSG00000025903.15   \n",
       "\n",
       "             transcript_id  \n",
       "361   ENSMUST00000134384.8  \n",
       "386  ENSMUST00000027036.11  \n",
       "409   ENSMUST00000150971.8  \n",
       "430   ENSMUST00000119612.9  \n",
       "449   ENSMUST00000137887.8  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gtfExonData = gtfPlusData.loc[gtfPlusData.feature == 'exon']\n",
    "gtfExonData[\"transcript_id\"] = gtfExonData.iloc[:,8].map(lambda p: p.split(';')[1].split(' ')[2][1:-1])\n",
    "gtfStartData = gtfPlusData.loc[gtfPlusData.feature == 'start_codon']\n",
    "gtfStartData[\"transcript_id\"] = gtfStartData.iloc[:,8].map(lambda p: p.split(';')[1].split(' ')[2][1:-1])\n",
    "gtfStopData = gtfPlusData.loc[gtfPlusData.feature == 'stop_codon']\n",
    "gtfStopData[\"transcript_id\"] = gtfStopData.iloc[:,8].map(lambda p: p.split(';')[1].split(' ')[2][1:-1])\n",
    "del gtfPlusData\n",
    "gtfStartData.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "circular-birmingham",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "10000\n",
      "20000\n",
      "30000\n",
      "40000\n",
      "50000\n",
      "60000\n",
      "70000\n",
      "80000\n",
      "90000\n",
      "100000\n"
     ]
    }
   ],
   "source": [
    "count = 0\n",
    "for ts in transDict.keys():\n",
    "    count += 1\n",
    "    if count % 10000 == 0:\n",
    "        print(count)\n",
    "    tsGtfExon = gtfExonData.loc[gtfExonData.transcript_id == ts]\n",
    "    tsGtfStart = gtfStartData.loc[gtfStartData.transcript_id == ts]\n",
    "    tsGtfStop = gtfStopData.loc[gtfStopData.transcript_id == ts]\n",
    "    for idx, row in tsGtfExon.iterrows():\n",
    "        transDict[ts][0].append(row['start'])\n",
    "        transDict[ts][1].append(row['end'])\n",
    "    if not (tsGtfStart.empty):\n",
    "        transDict[ts].append(['start_codon',tsGtfStart.iloc[0,3]] )\n",
    "    if not (tsGtfStop.empty):\n",
    "        transDict[ts].append(['stop_codon',tsGtfStop.iloc[0,3]] )\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "toxic-promise",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "21605\n"
     ]
    }
   ],
   "source": [
    "codingTransDict = {}\n",
    "for i in transDict.keys():\n",
    "    if len(transDict[i]) == 4:\n",
    "        codingTransDict[i] = transDict[i]\n",
    "print(len(codingTransDict))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "simple-genetics",
   "metadata": {},
   "outputs": [],
   "source": [
    "for i in codingTransDict.keys():\n",
    "    exonStart = codingTransDict[i][0]\n",
    "    exonEnd = codingTransDict[i][1]\n",
    "    startCod = codingTransDict[i][2][1]\n",
    "    stopCod = codingTransDict[i][3][1]\n",
    "    exonLength = 0\n",
    "    startPos = 0\n",
    "    stopPos = 0\n",
    "    for j in range(len(exonStart)):\n",
    "        if exonStart[j] <= startCod < exonEnd[j]:\n",
    "            startPos = exonLength + startCod - exonStart[j] + 1\n",
    "            exonLength = exonLength + exonEnd[j] - exonStart[j] + 1\n",
    "        else:\n",
    "            exonLength = exonLength + exonEnd[j] - exonStart[j] + 1\n",
    "    exonLength = 0\n",
    "    for j in range(len(exonStart)):\n",
    "        if exonStart[j] <= stopCod < exonEnd[j]:\n",
    "            stopPos = exonLength + stopCod - exonStart[j] + 1\n",
    "            exonLength = exonLength + exonEnd[j] - exonStart[j] + 1\n",
    "        else:\n",
    "            exonLength = exonLength + exonEnd[j] - exonStart[j] + 1\n",
    "    codingTransDict[i].append(exonLength)\n",
    "    codingTransDict[i].append(startPos)\n",
    "    codingTransDict[i].append(stopPos)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "closed-complement",
   "metadata": {},
   "outputs": [],
   "source": [
    "codingDF = pd.DataFrame.from_dict(codingTransDict, orient = 'index')\n",
    "codingDF = codingDF.drop(columns=[0,1,2,3,4])\n",
    "codingDF.to_csv('CodingTranscript.csv', sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "solved-adventure",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
